torchtune/models/llama4/_parallelism.py

```diff
@@ -0,0 +1,183 @@
+                f"decoder.layers.{layer_id}.mlp": PrepareModuleInputOutput(
+                    input_layouts=(Shard(1),),
+                    desired_input_layouts=(Replicate(),),
+                    use_local_input=True,
+                    output_layouts=(Partial(),),
+                    desired_output_layouts=(Shard(1),),
+                ),
+                # replicate computation for the router
+                f"decoder.layers.{layer_id}.mlp.router.gate": NoParallel(),
+                # input Replicate, output Partial
+                f"decoder.layers.{layer_id}.mlp.experts": ExpertTensorParallel(),
+            }
+        else:
+            mlp_plan = {
+                f"decoder.layers.{layer_id}.mlp": PrepareModuleInput(
+                    input_layouts=(Shard(1),),
+                    desired_input_layouts=(Replicate(),),
+                ),
+                f"decoder.layers.{layer_id}.mlp.w1": ColwiseParallel(),
+                f"decoder.layers.{layer_id}.mlp.w2": RowwiseParallel(
+                    output_layouts=Shard(1)
+                ),
+                f"decoder.layers.{layer_id}.mlp.w3": ColwiseParallel(),
+            }
+        layer_plan.update(mlp_plan)
+        plan.update(layer_plan)
+
+    return plan
+
+
+def decoder_only_tp_inference_plan(model: nn.Module) -> Dict[str, ParallelStyle]:
+    """
+    Helper function to get the tensor parallel plan for Llama4 where only the decoder is parallelized.
+    Usage of SequenceParallel requires that tp_dim % seq_len == 0, which will not hold in general
+    during autoregressive generation. As a result, we define this plan specifically for usage during generation.
+
+    Args:
+        model (nn.Module): Model to generate plan for
+
+    Returns:
+        Dict[str, Any]: The tensor parallel plan for Llama4 model.
+    """
+    plan = {
+        "decoder.output": ColwiseParallel(output_layouts=Replicate()),
+    }
+
+    for layer_id, transformer_block in enumerate(model.decoder.layers):
+
+        layer_plan = {
+            f"decoder.layers.{layer_id}.attn.q_proj": ColwiseParallel(),
+            f"decoder.layers.{layer_id}.attn.k_proj": ColwiseParallel(),
+            f"decoder.layers.{layer_id}.attn.v_proj": ColwiseParallel(),
+            f"decoder.layers.{layer_id}.attn.output_proj": RowwiseParallel(),
+            f"decoder.layers.{layer_id}.mlp_norm": SequenceParallel(),
+        }
+        if isinstance(transformer_block.mlp, MoE):
+            mlp_plan = {
+                # input / output sharding on the seqlen dim
+                # all-gather for input, reduce-scatter for output
+                f"decoder.layers.{layer_id}.mlp": PrepareModuleInputOutput(
+                    input_layouts=(Shard(1),),
+                    desired_input_layouts=(Replicate(),),
+                    use_local_input=True,
+                    output_layouts=(Partial(),),
+                    desired_output_layouts=(Shard(1),),
+                ),
```

Comment:
should be
```suggestion
                f"decoder.layers.{layer_id}.mlp": PrepareModuleOutput(
                    output_layouts=(Partial(),),
                    desired_output_layouts=(Replicate(),),
                ),
```
if you are not using sequence parallel for inference

Reply:
Tried this and it errored out, but need to debug a bit further